src/module.c

```diff
@@ -503,31 +503,38 @@ void *RM_PoolAlloc(RedisModuleCtx *ctx, size_t bytes) {
  * -------------------------------------------------------------------------- */
 
 client *moduleAllocTempClient() {
-    client *c;
+    client *c = NULL;
 
     pthread_mutex_lock(&moduleTempClientsMutex);
-    if (moduleTempClientCount > 0) {
+    if (moduleTempClientCount > 0)
         c = moduleTempClients[--moduleTempClientCount];
-    } else {
+    pthread_mutex_unlock(&moduleTempClientsMutex);
+
+    if (c == NULL) {
         c = createClient(NULL);
         c->flags |= CLIENT_MODULE;
         c->user = NULL; /* Root user */
     }
-    pthread_mutex_unlock(&moduleTempClientsMutex);
     return c;
 }
 
 void moduleReleaseTempClient(client *c) {
+    int full;
+
+    /* clearClient()/freeClient() may involve calls back to modules and they
+     * might also try to allocate temporary clients via moduleAllocTempClient().
+     * To avoid deadlock, we must be careful not to call these functions when we
+     * are holding the `moduleTempClientsMutex` lock. */
+    clearClient(c);
+    c->flags |= CLIENT_MODULE;
+    c->user = NULL; /* Root user */
```

Comment:
It seems that `clearClient` only needs when !full.

Reply:
Yes but we don't want to call clearClient/freeClient while holding the lock.(Please see Yossi's [comment](https://github.com/redis/redis/pull/9940#discussion_r772570598)). So, to detect if it's `full`, we have to acquire the lock. Then, we need to release the lock before calling `clearClient()`. After calling `clearClient()`, we need to acquire the lock again and place the client object into the cache if it has not become full while we are running `clearClient()`. 

So, It's a bit ugly but here we are being optimistic and calling `clearClient()` beforehand hoping cache is not full. In this case, we lock once. If cache is full (unlikely), yes, we are calling clearClient() unnecessarily.

Reply:
@tezc We can move it to the end of the method(`else`).

Reply:
@tezc Ohh, I missed something, you are right.

Reply:
@sundb  Could you please suggest as change here? 

If you are suggesting this : 

```c
if (full) {
    freeClient(c);
} else {
    clearClient(c);
}
```

At this point, we've already placed client into the cache and unlocked the mutex. Some other thread can acquire the lock and pull the client before we call `clearClient()`.

Reply:
@tezc Yeah, I see it, thx.